### PR TITLE
feat: Creates a modal that adds the ability to create a new custom attribute

### LINKED ISSUE #2866: feat: Add the ability to create a new custom attribute
- Start once issue #2864  is complete
- On clicking the "Add attribute" button on the attributes page, show a modal for entering the new attribute details. 
- Add the ability create a new attribute using `attribute_display_name`,  `attribute_description`, `attribute_display_type` (number, text, currency, percent, link.),  `attribute_model`(conversation_attribute or contact_attribute). 
- The `attribute_key` are unique and case sensitive, and can’t contain periods ('.'), dollar signs ('$'), characters like ~`!@#%^&*'{}|\'" or the NULL character. — If an unsupported character is used, the attribute will be created with an underscore in its place. show this item above the create button
- Show `attribute_key` above create button

**Design mock**
Screenshot 2021-08-20 at 5 18 37 PM

### PR DESCRIPTION
**Pull request by iamsivin**
# Pull Request Template

## Description

Add the ability to create a new custom attribute

Fixes #2866 

## Type of change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)

**Screen Recording**

https://user-images.githubusercontent.com/64252451/131292673-5f458098-b61c-4e21-b53c-b8038381b573.mov

## How Has This Been Tested?

Please import `import attributes from './attributes/attributes.routes';` and add `...attributes.routes,` in `settings.routes.js` file.

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: app/javascript/dashboard/routes/dashboard/settings/attributes/AddAttributes.vue [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+              <option v-for="type in types" :key="type.id" :value="type.id">
+                {{ type.option }}
+              </option>
+            </select>
+            <span v-if="$v.attributeType.$error" class="message">
+              {{ $t('ATTRIBUTES_MGMT.ADD.FORM.TYPE.ERROR') }}
+            </span>
+          </label>
+        </div>
+        <div v-if="hasValue" class="medium-12 columns">
```
Comment (muhsin-k) [inline app/javascript/dashboard/routes/dashboard/settings/attributes/AddAttributes.vue:56]:
```suggestion
        <div v-if="displayName" class="medium-12 columns">
```

File: app/javascript/dashboard/routes/dashboard/settings/attributes/AddAttributes.vue [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+          option: 'Currency',
+        },
+        {
+          id: 3,
+          option: 'Percent',
+        },
+        {
+          id: 4,
+          option: 'Link',
+        },
```
Comment (muhsin-k) [inline app/javascript/dashboard/routes/dashboard/settings/attributes/AddAttributes.vue:132]: Add `date`
```
{
  id: 5,
  option: "Date"
}
```

File: app/javascript/dashboard/routes/dashboard/settings/attributes/AddAttributes.vue [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+      ],
+      show: true,
+    };
+  },
+  computed: {
+    ...mapGetters({
+      uiFlags: 'getUIFlags',
+    }),
+    convertToSlug() {
```
Comment (muhsin-k) [inline app/javascript/dashboard/routes/dashboard/settings/attributes/AddAttributes.vue:141]:
```suggestion
    attributeKey() {
```

File: app/javascript/dashboard/routes/dashboard/settings/attributes/AddAttributes.vue [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+      return this.displayName
+        .toLowerCase()
+        .replace(/[^\w ]+/g, '')
+        .replace(/ +/g, '_');
+    },
+    hasValue() {
```
Comment (muhsin-k) [inline app/javascript/dashboard/routes/dashboard/settings/attributes/AddAttributes.vue:147]: I think this computed property is not needed. Just add a condition `displayName` is exist or not in the template

File: app/javascript/dashboard/routes/dashboard/settings/attributes/AddAttributes.vue [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+      required,
+    },
+    attributeType: {
+      required,
+    },
+  },
+
+  methods: {
+    showAlert(message) {
+      bus.$emit('newToastMessage', message);
```
Comment (muhsin-k) [inline app/javascript/dashboard/routes/dashboard/settings/attributes/AddAttributes.vue:169]: use `alertMixin`

File: app/javascript/dashboard/routes/dashboard/settings/attributes/AddAttributes.vue [hunk 6]
```diff
@@ -0,0 +1,194 @@
...
+    async addAttributes() {
+      try {
+        await this.$store.dispatch('attributes/create', {
+          attribute_display_name: this.displayName,
+          attribute_description: this.description,
+          attribute_model: this.attributeModel,
+          attribute_display_type: this.attributeType,
+          attribute_key: this.convertToSlug,
+        });
+        this.showAlert(this.$t('ATTRIBUTES_MGMT.ADD.API.SUCCESS_MESSAGE'));
```
Comment (muhsin-k) [inline app/javascript/dashboard/routes/dashboard/settings/attributes/AddAttributes.vue:180]: Show all messages in `finally` block
For ex:
```
 try {
 	await this.$store.dispatch('attributes/create', {
 		attribute_display_name: this.displayName,
 		attribute_description: this.description,
 		attribute_model: this.attributeModel,
 		attribute_display_type: this.attributeType,
 		attribute_key: this.convertToSlug,
 	});
 	this.alertMessage = this.$t('ATTRIBUTES_MGMT.ADD.API.SUCCESS_MESSAGE');
 	this.onClose();
 } catch (error) {
 	const errorMessage = error?.response?.data?.message;
 	this.alertMessage =
 		errorMessage || this.$t('ATTRIBUTES_MGMT.ADD.API.ERROR_MESSAGE');
 } finally {
 	this.showAlert(this.alertMessage);
 }
```

File: app/javascript/dashboard/routes/dashboard/settings/attributes/AddAttributes.vue [hunk 7]
```diff
@@ -0,0 +1,194 @@
...
+        this.onClose();
+      } catch (error) {
```
Comment (muhsin-k) [inline app/javascript/dashboard/routes/dashboard/settings/attributes/AddAttributes.vue:182]: Check any error message that exists in the `error` object otherways show the default error message.

File: app/javascript/dashboard/routes/dashboard/settings/attributes/Index.vue [hunk 8]
```diff
@@ -4,14 +4,36 @@
...
     </woot-button>
+    <woot-modal :show.sync="showAddPopup" :on-close="hideAddPopup">
+      <add-attributes :on-close="hideAddPopup" />
+    </woot-modal>
   </div>
 </template>
 
 <script>
-export default {};
+import AddAttributes from './AddAttributes';
```
Comment (muhsin-k) [inline app/javascript/dashboard/routes/dashboard/settings/attributes/Index.vue:18]:
```suggestion
import AddAttribute from './AddAttribute';
```

File: app/javascript/dashboard/store/mutation-types.js [hunk 9]
```diff
@@ -172,4 +172,11 @@ export default {
   SET_CSAT_RESPONSE_UI_FLAG: 'SET_CSAT_RESPONSE_UI_FLAG',
   SET_CSAT_RESPONSE: 'SET_CSAT_RESPONSE',
   SET_CSAT_RESPONSE_METRICS: 'SET_CSAT_RESPONSE_METRICS',
+
+  // Custom Attributes
+  SET_CUSTOM_ATTRIBUTES_UI_FLAG: 'SET_CUSTOM_ATTRIBUTES_UI_FLAG',
+  SET_CUSTOM_ATTRIBUTES: 'SET_CUSTOM_ATTRIBUTES',
```
Comment (muhsin-k) [inline app/javascript/dashboard/store/mutation-types.js:178]:
```suggestion
  SET_CUSTOM_ATTRIBUTE: 'SET_CUSTOM_ATTRIBUTE',
```

**Review by muhsin-k — CHANGES_REQUESTED** [review]
@iamsivin Nice work. Please see the inline comments

**Review by muhsin-k — APPROVED** [review]

**Review by nithindavid — APPROVED** [review]

**Pull request merged**